server/routes/api/protein.py

```diff
@@ -25,10 +30,293 @@
+    3) If an interaction ID is in scores but its reversal is not, D[interaction ID] = scores[interaction ID]
+    4) If both an interaction ID F and its reversal R are in scores, D[F] = D[R] = max(scores[F], scores[R])
+    '''
+    scores_sym = {}
+    for interaction_id, score in scores.items():
+        reverse_id = reverse_interaction_id(interaction_id)
+        reverse_score = scores.get(reverse_id, DEFAULT_INTERACTION_SCORE)
+        scores_sym[interaction_id] = scores_sym[reverse_id] = max(
+            score, reverse_score)
+    return scores_sym
+
+
+def flatten(nested_list):
+    '''
+    Given a list of lists, return a depth-1 flattened version
+    '''
+    flat = []
+    for L in nested_list:
+        flat.extend(L)
+    return flat
+
+
+def partition_expansion_cross(target_ids, node_set):
+    '''
+    Return two lists expansion_targets, cross_targets such that
+    1) expansion_targets U cross_targets = target_ids
+    2) expansion_targets and cross_targets are disjoint
+    3) all elements of cross_targets are in node_set
+    4) (implied) no elements of expansion_targets are in node_set
+    5) Stability: If id1 appears before id2 in target_ids and they are both assigned to the same
+       partition (either expansion_targets or cross_targets), id1 appears before id2 in this partition too.
+    '''
+    expansion_targets = []
+    cross_targets = []
+    for target_id in target_ids:
+        if target_id in node_set:
+            cross_targets.append(target_id)
+        else:
+            expansion_targets.append(target_id)
+    return expansion_targets, cross_targets
+
+
+@bp.route('/ppi/bfs/', methods=['POST'])
+def bfs():
+    '''
+    Retrieves graph data for protein-protein interaction graph via degree-bounded BFS
+
+    Request params:
+    - proteinDcid: the DCID of the page protein (e.g. 'bio/P53_HUMAN')
+    - scoreThreshold: interaction score threshold above which to display an edge between two interacting proteins
+    - numInteractors: we take {numInteractors} top-scoring expansion links of each node during BFS
+    - depth: the maximum distance of any returned node from the center protein
+
+    Returns: A dict of form
+    {
+        'nodeDataNested': [[...depth 0 node dicts], [...depth 1 node dicts], ...]
+        'linkDataNested': [[...depth 0 link dicts], [...depth 1 link dicts], ...]
+    }
+    The only depth 0 node is the center protein and since we do not support self-interactions at the moment,
+    there are no depth-0 links, so the above can be simplified to
+    {
+        'nodeDataNested': [[center protein dict], [...center protein neighbor dicts], ...]
+        'linkDataNested': [[], [...center protein link dicts], ...]
+    }.
+
+    TODO: example
+    '''
+    # adjacency list representation. maps interaction ids to list of target nodes, sorted in descending order by score
+    scores = {}
+
+    center_protein_dcid = escape(request.json['proteinDcid'])
```

Comment:
what if any of these request params are not present?

Reply:
Good point! should we respond with some sort of error here?